### DISCUSSION TITLE
UE5 UEReferenceApp build startup project fail (SOLVED)

### DISCUSSION THREAD

**Discussion by Mark-Moulton** [JSBSim-Team/jsbsim#623] [Support requests]
Hey all, I'd really appreciate some help, I'm trying to tinker with this in Unreal and running into some setup issues. Following the readme, I open and build the JSBSimForUnreal.sln with no errors, right click UEReferenceApp.uproject -> Generate visual studio project files, open up the solution and try to Build Startup Project on the UnrealVS extension bar. This is the output I get:
```
Build started...
1>------ Build started: Project: UEReferenceApp, Configuration: DebugGame_Editor x64 ------
1>Using bundled DotNet SDK
1>Log file: C:\Users\rando\AppData\Local\UnrealBuildTool\Log.txt
1>Creating makefile for UEReferenceAppEditor (no existing makefile)
1>UnrealBuildTool : warning : Library 'C:\jsbsim-master\UnrealEngine\Plugins\JSBSimFlightDynamicsModel\Source\ThirdParty\JSBSim\Lib\JSBSim.lib' was not resolvable to a file when used in Module 'JSBSim', assuming it is a filename and will search library paths for it. This is slow and dependency checking will not work for it. Please update reference to be fully qualified alternatively use PublicSystemLibraryPaths if you do intended to use this slow path to suppress this warning.
1>Parsing headers for UEReferenceAppEditor
1>  Running UnrealHeaderTool "C:\jsbsim-master\UnrealEngine\UEReferenceApp.uproject" "C:\jsbsim-master\UnrealEngine\Intermediate\Build\Win64\UEReferenceAppEditor\DebugGame\UEReferenceAppEditor.uhtmanifest" -LogCmds="loginit warning, logexit warning, logdatabase error" -Unattended -WarningsAsErrors -abslog="C:\Users\rando\AppData\Local\UnrealBuildTool\Log_UHT.txt" -installed
1>Reflection code generated for UEReferenceAppEditor in 1.5128132 seconds
1>Building UEReferenceAppEditor...
1>Using Visual Studio 2019 14.29.30145 toolchain (C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133) and Windows 10.0.18362.0 SDK (C:\Program Files (x86)\Windows Kits\10).
1>[Adaptive Build] Excluded from UEReferenceApp unity file: UEReferenceApp.cpp, UEReferenceAppGameModeBase.cpp
1>Determining max actions to execute in parallel (8 physical cores, 16 logical cores)
1>  Executing up to 8 processes, one per physical core
1>Building 19 actions with 8 processes...
1>[1/19] Resource Default.rc2
1>[2/19] Resource Default.rc2
1>[3/19] Resource Default.rc2
1>[4/19] Compile SharedPCH.UnrealEd.NonOptimized.ShadowErrors.cpp
1>[5/19] Compile SharedPCH.Engine.NonOptimized.ShadowErrors.cpp
1>[6/19] Compile UEReferenceApp.init.gen.cpp
1>[7/19] Compile UEReferenceApp.cpp
1>[8/19] Compile UEReferenceAppGameModeBase.gen.cpp
1>[9/19] Compile UEReferenceAppGameModeBase.cpp
1>[10/19] Link UnrealEditor-UEReferenceApp-Win64-DebugGame.lib
1>   Creating library C:\jsbsim-master\UnrealEngine\Intermediate\Build\Win64\UnrealEditor\DebugGame\UEReferenceApp\UnrealEditor-UEReferenceApp-Win64-DebugGame.lib and object C:\jsbsim-master\UnrealEngine\Intermediate\Build\Win64\UnrealEditor\DebugGame\UEReferenceApp\UnrealEditor-UEReferenceApp-Win64-DebugGame.exp
1>[11/19] Link UnrealEditor-UEReferenceApp-Win64-DebugGame.dll
1>   Creating library C:\jsbsim-master\UnrealEngine\Intermediate\Build\Win64\UnrealEditor\DebugGame\UEReferenceApp\UnrealEditor-UEReferenceApp-Win64-DebugGame.suppressed.lib and object C:\jsbsim-master\UnrealEngine\Intermediate\Build\Win64\UnrealEditor\DebugGame\UEReferenceApp\UnrealEditor-UEReferenceApp-Win64-DebugGame.suppressed.exp
1>[12/19] Compile Module.JSBSimFlightDynamicsModel.gen.cpp
1>[13/19] Compile Module.JSBSimFlightDynamicsModel.cpp
1>[14/19] Link UnrealEditor-JSBSimFlightDynamicsModel-Win64-DebugGame.lib
1>   Creating library C:\jsbsim-master\UnrealEngine\Plugins\JSBSimFlightDynamicsModel\Intermediate\Build\Win64\UnrealEditor\DebugGame\JSBSimFlightDynamicsModel\UnrealEditor-JSBSimFlightDynamicsModel-Win64-DebugGame.lib and object C:\jsbsim-master\UnrealEngine\Plugins\JSBSimFlightDynamicsModel\Intermediate\Build\Win64\UnrealEditor\DebugGame\JSBSimFlightDynamicsModel\UnrealEditor-JSBSimFlightDynamicsModel-Win64-DebugGame.exp
1>[15/19] Link UnrealEditor-JSBSimFlightDynamicsModel-Win64-DebugGame.dll
1>LINK : fatal error LNK1181: cannot open input file 'C:\jsbsim-master\UnrealEngine\Plugins\JSBSimFlightDynamicsModel\Source\ThirdParty\JSBSim\Lib\JSBSim.lib'
1>[16/19] Compile Module.JSBSimFlightDynamicsModelEditor.cpp
1>[17/19] Link UnrealEditor-JSBSimFlightDynamicsModelEditor-Win64-DebugGame.lib
1>   Creating library C:\jsbsim-master\UnrealEngine\Plugins\JSBSimFlightDynamicsModel\Intermediate\Build\Win64\UnrealEditor\DebugGame\JSBSimFlightDynamicsModelEditor\UnrealEditor-JSBSimFlightDynamicsModelEditor-Win64-DebugGame.lib and object C:\jsbsim-master\UnrealEngine\Plugins\JSBSimFlightDynamicsModel\Intermediate\Build\Win64\UnrealEditor\DebugGame\JSBSimFlightDynamicsModelEditor\UnrealEditor-JSBSimFlightDynamicsModelEditor-Win64-DebugGame.exp
1>[18/19] Link UnrealEditor-JSBSimFlightDynamicsModelEditor-Win64-DebugGame.dll
1>LINK : fatal error LNK1181: cannot open input file 'C:\jsbsim-master\UnrealEngine\Plugins\JSBSimFlightDynamicsModel\Source\ThirdParty\JSBSim\Lib\JSBSim.lib'
1>C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.MakeFile.Targets(44,5): error MSB3073: The command ""C:\Program Files\Epic Games\UE_5.0\Engine\Build\BatchFiles\Build.bat" UEReferenceAppEditor Win64 DebugGame -Project="C:\jsbsim-master\UnrealEngine\UEReferenceApp.uproject" -WaitMutex -FromMsBuild" exited with code 6.
1>Done building project "UEReferenceApp.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```
![image](https://user-images.githubusercontent.com/8378246/168445618-99f89b73-5424-430e-8212-b0b33a105b2a.png)

edit: Found my error. Paying a little closer attention I noticed it was creating a "LibDebug" folder instead of "Lib". I realized I wasn't building the JSBSimForUnreal.sln in RELEASE mode.

I'm embarrassed, hope this helps someone in the future.

**Comment by bcoconni** [comment]
According to [the docs](https://github.com/JSBSim-Team/jsbsim/blob/27bf560a3ffe10a4fb146803a5d931a9e1f3c492/UnrealEngine/README.md), it seems that JSBSim must be built separately as a static library and copied to the folder `Plugins/JSBSimFlightDynamicsModel`:

> **2. Build JSBSim as Static libraries and stage Model files**
>
> Unreal Engine requires that one plugin contains all its needed files in its sub-folders. 
This application contains a `Plugins/JSBSimFlightDynamicsModel` folder containing the wrapper around JSBSim.
> In some of these subfolders, one has to place 
> - The JSBSim libraries, compiled as static libs  
> - The aircrafts/engine/systems definition files.

More specifically and according to the error message, the linker is expecting `JSBSim.lib` to be located in the folder `Plugins\JSBSimFlightDynamicsModel\Source\ThirdParty\JSBSim\Lib`.

You can use the visual studio project file `JSBSim.sln` located at the root folder of JSBSim to build the JSBSim library.

May be @AlbanBERGERET-Epic can provide more details or correct me if I'm wrong ?

**Reply by Mark-Moulton** [reply]
> 2. Build JSBSim as Static libraries and stage Model files
> 
> Unreal Engine requires that one plugin contains all its needed files in its sub-folders. This application contains a Plugins/JSBSimFlightDynamicsModel folder containing the wrapper around JSBSim. In some of these subfolders, one has to place
> 
>- The JSBSim libraries, compiled as static libs
>- The aircrafts/engine/systems definition files.
> 
> When the UE application will be packaged, the resources will be copied along with the executable, and the application linked against the static libs transparently.
> 
> To make this process easier, there is a new solution named JSBSimForUnreal.sln at the root of JSBSim repo.
> 
> - Simply open and build this solution with VS2019, in Release, (and in Debug if you want too, but this is not mandatory)
> - It will take care of making a clean build, and copy all needed files at the right location
>   - All libs and headers in UnrealEngine\Plugins\JSBSimFlightDynamicsModel\Source\ThirdParty\JSBSim
>   - All resource files (aircrafts/engines/systems) in UnrealEngine\Plugins\JSBSimFlightDynamicsModel\Resources\JSBSim

I was using the second part with JSBSimForUnreal.sln. Worth a shot trying it manually though, I'll take another look tomorrow when I get a chance.

edit: Found my error. Paying a little closer attention I noticed it was creating a "LibDebug" folder instead of "Lib". I realized I wasn't building the JSBSimForUnreal.sln in RELEASE mode.

I'm embarrassed, hope this helps someone in the future.

**Reply by AlbanBERGERET-Epic** [reply]
Hi @NPCsoft 

Yes, @bcoconni, you are right : The error messages in the build are typically those of a missing lib file. 
`1>LINK : fatal error LNK1181: cannot open input file 'C:\jsbsim-master\UnrealEngine\Plugins\JSBSimFlightDynamicsModel\Source\ThirdParty\JSBSim\Lib\JSBSim.lib'`

Glad that you found the error. It also could have come from the Configuration you choose for the UE Reference Application. 

For further reference, yes, the JSBSimForUnreal.sln comes with two different targets - Release and Debug.
For 99% of the cases, just building the Release one is needed - Unreal will still allow you to debug your plugin and application with Release third parties. The debug is really if you want to have symbols of the JSBSim classes to debug JSBSim itself... 

UE Application Configuration
DebugGame                   -->  Uses Debug JSBSim Libs
DebugGame Editor        -->  Uses Debug JSBSim Libs
Development Game       -->  Uses Release JSBSim Libs
Development Editor       -->  Uses Release JSBSim Libs
Shipping                         -->  Uses Release JSBSim Libs

I hope it helps. 

Alban
